File: exercises/python/order_service/application/checkout_saga.py
"""Checkout saga -- the application use case that orchestrates calls
across the Inventory, Payment, and Shipping bounded contexts and emits
the resulting domain events to Notification (via Kafka).

Architecture
~~~~~~~~~~~~
The saga depends only on domain types and outbound ports. It has no
awareness of REST, gRPC, JSON, Kafka, or any specific transport. Adapter
implementations of those ports live in ``infrastructure/``.

Observability instrumentation reference
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
This class is the canonical example for Modules 3a, 3b, 3c:

- **Module 3a** (structured logs): ``DomainContext`` populates structlog
  context with ``order.id``, ``customer.id``, ``cart.id``.
- **Module 3b** (domain spans + business attributes): ``@WithSpan``-style
  span naming + ``set_baggage("customer.tier")`` propagation.
- **Module 3c** (business metrics): ``checkout_outcomes_total`` counter
  and ``checkout_duration_seconds`` histogram.
"""

from __future__ import annotations

import time
from dataclasses import dataclass

import structlog
from opentelemetry import metrics, trace

from domain.events import OrderCancelled, OrderConfirmed, OrderPlaced
from domain.identifiers import OrderContextKey
from domain.models import Order, OrderId, OrderStatus
from domain.ports import (
    Authorized,
    AuthorizationFailure,
    AuthorizationOutcome,
    Declined,
    InventoryPort,
    OrderEventPublisher,
    PaymentPort,
    ReservationFailure,
    ReservationOutcome,
    Reserved,
    Scheduled,
    ShipmentFailure,
    ShipmentOutcome,
    ShippingPort,
    Unavailable,
)
from domain.services import CustomerProfile, CustomerProfileLookup
from shared_observability import DomainContext, set_baggage

logger = structlog.get_logger()
tracer = trace.get_tracer("order-service")
meter = metrics.get_meter("order-service")

# Module 3c: business metrics
checkout_outcomes_counter = meter.create_counter(
    "checkout_outcomes_total",
    description="Total checkout outcomes by result and customer tier",
)

checkout_duration_histogram = meter.create_histogram(
    "checkout_duration_seconds",
    description="End-to-end checkout saga duration",
    unit="s",
)


# ---------------------------------------------------------------------------
# Command & Result
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class CheckoutCommand:
    """Input to the checkout saga use case.

    Plain dataclass, no framework dependencies. Inputs are expressed in
    domain types -- the conversion from inbound HTTP DTOs lives in the
    web adapter.
    """

    customer_id: str
    cart_id: str
    line_items: list  # list[LineItem] -- kept untyped to avoid circular import
    payment_method: str = "credit_card"
    shipping_class: str = "standard"


@dataclass(frozen=True)
class CheckoutResult:
    """Output of the checkout saga.

    A single dataclass covers both confirmed and cancelled outcomes.
    The ``status`` field discriminates between them.
    """

    order_id: str
    status: str  # "CONFIRMED" or "CANCELLED"
    reservation_id: str | None = None
    authorization_id: str | None = None
    shipment_id: str | None = None
    message: str = ""


# ---------------------------------------------------------------------------
# Saga
# ---------------------------------------------------------------------------


class CheckoutSaga:
    """Orchestration-based saga for the checkout flow.

    Coordinates Inventory -> Payment -> Shipping, publishing domain events
    at each milestone. If any step fails, the order is cancelled and an
    ``OrderCancelled`` event is published.
    """

    def __init__(
        self,
        inventory: InventoryPort,
        payment: PaymentPort,
        shipping: ShippingPort,
        events: OrderEventPublisher,
        customer_lookup: CustomerProfileLookup,
    ) -> None:
        self._inventory = inventory
        self._payment = payment
        self._shipping = shipping
        self._events = events
        self._customer_lookup = customer_lookup

    def checkout(self, command: CheckoutCommand) -> CheckoutResult:
        """Execute the checkout saga.

        Returns a CheckoutResult that the caller (typically a FastAPI
        route) translates into an HTTP response.
        """
        from domain.models import CartId, CustomerId, LineItem, Money, Sku

        order_id = OrderId.generate()

        # Module 3a: populate structlog context with domain identifiers
        with DomainContext(
            OrderContextKey.ORDER_ID.of(order_id.value),
            OrderContextKey.CUSTOMER_ID.of(command.customer_id),
            OrderContextKey.CART_ID.of(command.cart_id),
        ):
            # Build domain objects from command
            customer_id = CustomerId.of(command.customer_id)
            cart_id = CartId.of(command.cart_id)
            line_items = command.line_items

            # Place the order in the domain
            order = Order.place(order_id, customer_id, cart_id, line_items)

            # Look up the customer's tier
            profile = self._customer_lookup.lookup(customer_id)

            # Module 3b: annotate this span with business attributes
            with tracer.start_as_current_span("Order.Checkout") as span:
                span.set_attribute("order.id", order.id.value)
                span.set_attribute("order.value", float(order.total().amount))
                span.set_attribute(
                    "order.line_items_count", len(order.line_items)
                )
                span.set_attribute("customer.id", order.customer_id.value)
                span.set_attribute("customer.tier", profile.tier.value)

                logger.info(
                    "Checkout starting",
                    total=str(order.total().amount),
                    items=len(order.line_items),
                )

                # Module 3b: propagate customer.tier downstream as baggage
                set_baggage("customer.tier", profile.tier.value)

                return self._run_saga(order, profile)

    # ------------------------------------------------------------------------
    # Saga steps
    # ------------------------------------------------------------------------

    def _run_saga(
        self, order: Order, profile: CustomerProfile
    ) -> CheckoutResult:
        start_time = time.monotonic()
        try:
            # Step 1: publish OrderPlaced (always, regardless of saga outcome)
            self._publish_event(OrderPlaced.from_order(order))

            # Step 2: reserve stock
            reservation_outcome = self._inventory.reserve(order)

            if isinstance(reservation_outcome, Reserved):
                return self._continue_after_reservation(
                    order,
                    profile,
                    reservation_outcome.reservation_id,
                    start_time,
                )
            elif isinstance(reservation_outcome, Unavailable):
                return self._cancel_and_record(
                    order,
                    profile,
                    "inventory",
                    reservation_outcome.reason,
                    start_time,
                )
            else:  # ReservationFailure
                return self._cancel_and_record(
                    order,
                    profile,
                    "inventory",
                    f"inventory failure: {reservation_outcome.detail}",
                    start_time,
                )

        except Exception as e:
            logger.error("Saga aborted unexpectedly", error=str(e))
            return self._cancel_and_record(
                order, profile, "saga", str(e), start_time
            )

    def _continue_after_reservation(
        self,
        order: Order,
        profile: CustomerProfile,
        reservation_id: str,
        start_time: float,
    ) -> CheckoutResult:
        # Step 3: authorize payment
        auth_outcome = self._payment.authorize(order)

        if isinstance(auth_outcome, Authorized):
            return self._continue_after_authorization(
                order,
                profile,
                reservation_id,
                auth_outcome.authorization_id,
                start_time,
            )
        elif isinstance(auth_outcome, Declined):
            return self._cancel_and_record(
                order,
                profile,
                "payment",
                auth_outcome.reason,
                start_time,
            )
        else:  # AuthorizationFailure
            return self._cancel_and_record(
                order,
                profile,
                "payment",
                f"payment failure: {auth_outcome.detail}",
                start_time,
            )

    def _continue_after_authorization(
        self,
        order: Order,
        profile: CustomerProfile,
        reservation_id: str,
        authorization_id: str,
        start_time: float,
    ) -> CheckoutResult:
        # Step 4: schedule shipment
        shipping_outcome = self._shipping.schedule(order)

        if isinstance(shipping_outcome, Scheduled):
            confirmed = order.confirm()
            self._publish_event(
                OrderConfirmed.from_order(
                    confirmed,
                    reservation_id,
                    authorization_id,
                    shipping_outcome.shipment_id,
                )
            )
            self._record_outcome("success", profile, start_time)
            return CheckoutResult(
                order_id=confirmed.id.value,
                status="CONFIRMED",
                reservation_id=reservation_id,
                authorization_id=authorization_id,
                shipment_id=shipping_outcome.shipment_id,
                message="Order confirmed successfully",
            )
        else:  # ShipmentFailure
            return self._cancel_and_record(
                order,
                profile,
                "shipping",
                f"shipping failure: {shipping_outcome.detail}",
                start_time,
            )

    # ------------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------------

    def _cancel_and_record(
        self,
        order: Order,
        profile: CustomerProfile,
        failed_at: str,
        reason: str,
        start_time: float,
    ) -> CheckoutResult:
        cancelled = order.cancel(reason)
        self._publish_event(
            OrderCancelled.from_order(cancelled, failed_at, reason)
        )
        self._record_outcome(f"cancelled_{failed_at}", profile, start_time)
        logger.warning(
            "Checkout cancelled", failed_at=failed_at, reason=reason
        )
        return CheckoutResult(
            order_id=cancelled.id.value,
            status="CANCELLED",
            message=f"Cancelled at {failed_at}: {reason}",
        )

    def _publish_event(self, event) -> None:
        """Publish a domain event, swallowing failures to avoid aborting the saga."""
        try:
            self._events.publish(event)
        except Exception as e:
            # Event publication failure should not abort the saga
            logger.error(
                "Failed to publish event",
                event_type=event.event_type,
                error=str(e),
            )

    def _record_outcome(
        self, outcome: str, profile: CustomerProfile, start_time: float
    ) -> None:
        """Record per-outcome, per-tier counter and duration histogram."""
        duration = time.monotonic() - start_time
        checkout_outcomes_counter.add(
            1,
            {"outcome": outcome, "tier": profile.tier.value},
        )
        checkout_duration_histogram.record(
            duration,
            {"outcome": outcome, "tier": profile.tier.value},
        )

File: exercises/quarkus/order-service/src/main/java/com/example/order/application/CheckoutSaga.java
package com.example.order.application;

import com.example.order.domain.event.DomainEvent;
import com.example.order.domain.event.OrderCancelled;
import com.example.order.domain.event.OrderConfirmed;
import com.example.order.domain.event.OrderPlaced;
import com.example.order.domain.identifier.OrderContextKey;
import com.example.order.domain.model.Order;
import com.example.order.domain.model.OrderId;
import com.example.order.domain.outbound.AuthorizationOutcome;
import com.example.order.domain.outbound.InventoryPort;
import com.example.order.domain.outbound.OrderEventPublisher;
import com.example.order.domain.outbound.PaymentPort;
import com.example.order.domain.outbound.ReservationOutcome;
import com.example.order.domain.outbound.ShipmentOutcome;
import com.example.order.domain.outbound.ShippingPort;
import com.example.order.domain.service.CustomerProfile;
import com.example.order.domain.service.CustomerProfileLookup;
import com.example.workshop.observability.BaggageHelpers;
import com.example.workshop.observability.DomainContext;

import io.micrometer.core.instrument.MeterRegistry;
import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.metrics.DoubleHistogram;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.instrumentation.annotations.WithSpan;
import io.opentelemetry.context.Scope;
import java.util.List;
import jakarta.enterprise.context.ApplicationScoped;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.atomic.AtomicInteger;

/**
 * The checkout saga - the application use case that orchestrates calls
 * across the Inventory, Payment, and Shipping bounded contexts and emits
 * the resulting domain events to Notification (via Kafka).
 *
 * <h2>Architecture</h2>
 *
 * <p>The saga depends only on domain types and outbound ports. It has
 * no awareness of REST, gRPC, JSON, Kafka, or any specific transport.
 * Adapter implementations of those ports live in {@code infrastructure/}.
 *
 * <h2>Observability instrumentation reference</h2>
 *
 * <p>This class is the canonical example for Modules 3a, 3b, 3c. The
 * {@code main} branch ships it fully instrumented; the workshop's
 * retrograde branches strip pieces back so participants experience adding
 * each layer themselves. Specifically:
 *
 * <ul>
 *   <li><b>Module 3a (structured logs)</b>: the try-with-resources
 *       {@code DomainContext.open(...)} block at the top of
 *       {@code checkout()} populates SLF4J MDC with this context's
 *       identifiers. Every log line emitted in that scope - and
 *       transitively from the adapters this method calls - carries
 *       {@code order.id}, {@code customer.id}, {@code cart.id}.</li>
 *   <li><b>Module 3b (domain spans + business attributes)</b>: the
 *       {@code @WithSpan("Order.Checkout")} annotation names the saga's
 *       span in domain language. {@code Span.current().setAttribute(...)}
 *       calls add business-meaningful attributes ({@code order.value},
 *       {@code customer.tier}, {@code order.line_items_count}). The
 *       {@code BaggageHelpers.put(...)} block propagates
 *       {@code customer.tier} downstream so spans in other services
 *       can be tagged by tier without each service re-fetching the
 *       profile.</li>
 *   <li><b>Module 3c (business metrics)</b>: the three custom Micrometer
 *       meters - {@code checkout_orders_in_payment_verification} (gauge),
 *       {@code checkout_outcomes_total} (counter with {@code outcome}
 *       and {@code tier} labels), and {@code checkout_duration_seconds}
 *       (timer) - answer the three business questions from the module.</li>
 * </ul>
 */
@ApplicationScoped
public class CheckoutSaga {

    private static final Logger log = LoggerFactory.getLogger(CheckoutSaga.class);

    // -----------------------------------------------------------------------
    // Collaborators - all injected via constructor
    // -----------------------------------------------------------------------
    private final InventoryPort inventory;
    private final PaymentPort payment;
    private final ShippingPort shipping;
    private final OrderEventPublisher events;
    private final CustomerProfileLookup customerLookup;

    // -----------------------------------------------------------------------
    // Custom metrics - registered eagerly in the constructor
    // -----------------------------------------------------------------------
    private final AtomicInteger ordersInPaymentVerification = new AtomicInteger(0);
    private final MeterRegistry meterRegistry;
    private final DoubleHistogram checkoutDurationHistogram;

    public CheckoutSaga(InventoryPort inventory,
                        PaymentPort payment,
                        ShippingPort shipping,
                        OrderEventPublisher events,
                        CustomerProfileLookup customerLookup,
                        MeterRegistry meterRegistry) {
        this.inventory = inventory;
        this.payment = payment;
        this.shipping = shipping;
        this.events = events;
        this.customerLookup = customerLookup;
        this.meterRegistry = meterRegistry;

        // Module 3c: gauge for "orders currently in payment verification"
        meterRegistry.gauge(
                "checkout_orders_in_payment_verification",
                ordersInPaymentVerification,
                AtomicInteger::get);

        // Module 3c: end-to-end checkout duration, recorded via the OpenTelemetry
        // API directly (not a Micrometer Timer) so the exported name matches the
        // Python and C# services and the Grafana dashboards exactly:
        // workshop_checkout_duration_seconds{_bucket,_count,_sum}. A Micrometer
        // Timer would export a milliseconds unit and become
        // ..._seconds_milliseconds, which no dashboard queries.
        this.checkoutDurationHistogram = GlobalOpenTelemetry.get()
                .getMeter("order-service")
                .histogramBuilder("checkout_duration_seconds")
                .setDescription("End-to-end checkout saga duration")
                .setUnit("s")
                .setExplicitBucketBoundariesAdvice(List.of(
                        0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0))
                .build();
    }

    /**
     * Execute the checkout saga. Returns a {@link CheckoutResult} that the
     * caller (typically a JAX-RS resource) translates into a response.
     */
    @WithSpan("Order.Checkout")
    public CheckoutResult checkout(CheckoutCommand command) {
        var orderId = OrderId.generate();

        // Module 3a: populate MDC with this context's identifiers for the
        // duration of the saga. All log lines in this scope carry these.
        try (var ctx = DomainContext.open(
                OrderContextKey.ORDER_ID.of(orderId.value()),
                OrderContextKey.CUSTOMER_ID.of(command.customerId().value()),
                OrderContextKey.CART_ID.of(command.cartId().value()))) {

            // Place the order in the domain.
            var order = Order.place(orderId, command.customerId(),
                    command.cartId(), command.lineItems());

            // Look up the customer's tier - drives downstream baggage and
            // the per-tier metrics breakdown.
            CustomerProfile profile = customerLookup.lookup(command.customerId());

            // Module 3b: annotate this span with business attributes.
            Span span = Span.current();
            span.setAttribute("order.id", order.id().value());
            span.setAttribute("order.value", order.total().amount().doubleValue());
            span.setAttribute("order.line_items_count", order.lineItems().size());
            span.setAttribute("customer.id", order.customerId().value());
            span.setAttribute("customer.tier", profile.tier().name());

            log.info("Checkout starting: total={} items={}",
                    order.total().amount(), order.lineItems().size());

            // Module 3b: propagate customer.tier downstream as baggage so
            // every service in the saga can tag its own observability by
            // tier without re-fetching.
            try (Scope baggageScope = BaggageHelpers.put(
                    OrderContextKey.CUSTOMER_TIER.of(profile.tier().name()))) {

                return runSaga(order, profile);
            }
        }
    }

    // ------------------------------------------------------------------------
    // Saga steps
    // ------------------------------------------------------------------------

    private CheckoutResult runSaga(Order order, CustomerProfile profile) {
        long startNanos = System.nanoTime();
        try {
            // Step 1: place-event. Always published, regardless of saga outcome -
            // OrderPlaced records the customer's intent.
            publishEvent(OrderPlaced.from(order));

            // Step 2: reserve stock. Sealed-interface switch is exhaustive,
            // so every path produces exactly one CheckoutResult and exactly
            // one outcome metric increment.
            var reservationOutcome = inventory.reserve(order);
            return switch (reservationOutcome) {
                case ReservationOutcome.Reserved r ->
                        continueAfterReservation(order, profile, r.reservationId(), startNanos);
                case ReservationOutcome.Unavailable u ->
                        cancelAndRecord(order, profile, "inventory", u.reason(), startNanos);
                case ReservationOutcome.Failure f ->
                        cancelAndRecord(order, profile, "inventory",
                                "inventory failure: " + f.detail(), startNanos);
            };
        } catch (RuntimeException e) {
            log.error("Saga aborted unexpectedly", e);
            return cancelAndRecord(order, profile, "saga", e.getMessage(), startNanos);
        }
    }

    private CheckoutResult continueAfterReservation(Order order,
                                                    CustomerProfile profile,
                                                    String reservationId,
                                                    long startNanos) {
        // Step 3: authorize payment. Module 3c: gauge in/out around this call.
        ordersInPaymentVerification.incrementAndGet();
        AuthorizationOutcome authOutcome;
        try {
            authOutcome = payment.authorize(order);
        } finally {
            ordersInPaymentVerification.decrementAndGet();
        }

        return switch (authOutcome) {
            case AuthorizationOutcome.Authorized a ->
                    continueAfterAuthorization(order, profile, reservationId,
                            a.authorizationId(), startNanos);
            case AuthorizationOutcome.Declined d ->
                    cancelAndRecord(order, profile, "payment", d.reason(), startNanos);
            case AuthorizationOutcome.Failure f ->
                    cancelAndRecord(order, profile, "payment",
                            "payment failure: " + f.detail(), startNanos);
        };
    }

    private CheckoutResult continueAfterAuthorization(Order order,
                                                      CustomerProfile profile,
                                                      String reservationId,
                                                      String authorizationId,
                                                      long startNanos) {
        // Step 4: schedule shipment.
        var shippingOutcome = shipping.schedule(order);
        return switch (shippingOutcome) {
            case ShipmentOutcome.Scheduled s -> {
                var confirmed = order.confirm();
                publishEvent(OrderConfirmed.from(confirmed, reservationId,
                        authorizationId, s.shipmentId()));
                recordOutcome("success", profile, startNanos);
                yield new CheckoutResult.Confirmed(
                        confirmed.id(), reservationId, authorizationId, s.shipmentId());
            }
            case ShipmentOutcome.Failure f ->
                    cancelAndRecord(order, profile, "shipping",
                            "shipping failure: " + f.detail(), startNanos);
        };
    }

    // ------------------------------------------------------------------------
    // Helpers
    // ------------------------------------------------------------------------

    private CheckoutResult cancelAndRecord(Order order,
                                           CustomerProfile profile,
                                           String failedAt,
                                           String reason,
                                           long startNanos) {
        var cancelled = order.cancel(reason);
        publishEvent(OrderCancelled.from(cancelled, failedAt, reason));
        recordOutcome("cancelled_" + failedAt, profile, startNanos);
        log.warn("Checkout cancelled at {}: {}", failedAt, reason);
        return new CheckoutResult.Cancelled(cancelled.id(), failedAt, reason);
    }

    private void publishEvent(DomainEvent event) {
        try {
            events.publish(event);
        } catch (RuntimeException e) {
            // Event publication failure should not abort the saga - log loudly
            // and continue. A real system would also write the event to an
            // outbox table for retry.
            log.error("Failed to publish {}: {}", event.eventType(), e.getMessage(), e);
        }
    }

    private void recordOutcome(String outcome, CustomerProfile profile, long startNanos) {
        long durationNanos = System.nanoTime() - startNanos;
        // Module 3c: per-outcome, per-tier counter and timer.
        meterRegistry.counter("checkout_outcomes_total",
                "outcome", outcome,
                "tier", profile.tier().name()).increment();
        checkoutDurationHistogram.record(durationNanos / 1_000_000_000.0,
                Attributes.of(
                        AttributeKey.stringKey("outcome"), outcome,
                        AttributeKey.stringKey("tier"), profile.tier().name()));
    }
}
